docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2624 @@
+</details>
+
+<details>
+<summary>Option-2: Independent single node clusters with Mirror-Maker</summary>
+<img src="assets/activepassive-option2.png">
+
+An extension to `Option-1` where we do not only rely on snapshots (full/delta) to keep the `hot-standy` etcd node in sync with the `primary etcd`. etcd provides an alternative way via [mirror maker](https://github.com/etcd-io/etcd/blob/main/etcdctl/doc/mirror_maker.md) to continuously relay key creates and updates to a separate cluster. In this option reconciliation from snapshots in backup-bucket will be used as a fallback.
+
+**Pros**
+* All the pros in `Option-1` are also applicable for this option
+* Using mirror maker provides continuous synchronization of key creates/updates to the `hot-standby` etcd, making it less dependent on reconciliaton from snapshots fetched from the backup-bucket.
+
+**Cons**
+* All the cons in `Option-1` are also applicable for this option.
+* Mirroring does not guarantee linearizability. During network disconnects, previous key-values might have been discarded, and clients are expected to verify watch responses for correct ordering. Thus, there is no ordering guarantee in mirror. So the client wishes to preserve all historical data and ordering then a learner approach of synchronization is better as compared to using a mirror maker.
+</details>
+ 
+<details>
+<summary>Option-3: Perpetual Learner</summary>
+<img src="assets/activepassive-option3.png">
+
+In this option etcd cluster and learner facilities are leveraged. `etcd-druid` will bootstrap a cluster with one member. Once this member is ready to serve client requests then an additional `learner` will be added which joins the etcd cluster as a non-voting member. Learner will reject client reads and writes requests and the clients will have to reattempt. Typically switching to another member in a cluster post retries is provided out-of-the-box by etcd `clientv3`. The only `member` who is also the leader will serve all client requests.
+
+**Pros**
+* All the pros in `Option-1` are also applicable for this option.
+* Leaner will be continuously updated by the leader and will remain in-sync with the leader. Using a learner retains historical data and its ordering and is therefore better in that aspect as compared to `Option-2`.
+
+**Cons**
+* All the cons in `Option-1` are also applicable for this option.
+* `etcd-druid` will now have to additionaly play an active role in managing members of an etcd cluster by add new members as `learner` and promoting `learner` to an active member if the leader is no longer available. This will increase the complexity in `etcd-druid`.
+* To prevent clients from re-attempting to reach the `learner`, `etcd-druid` will have to ensure that the label on the learner are set differently than the `leader`. This needs to be done everytime there is a switch in the leader/learner.
```

Comment:
I guess, that’s what should prevent the disadvantage above, so it’s not really an actual (impactful) disadvantage (can be relatively easily avoided in Kubernetes).

Reply:
True, Cons as its extra work :)

Reply:
Thanks.